query-engine/schema/src/identifier_type.rs

```diff
@@ -0,0 +1,319 @@
+use crate::capitalize;
+
+use prisma_models::{prelude::*, *};
+
+/// Enum used to represent unique schema type names.
+/// It helps deferring the allocation + formatting of strings
+/// during the initialization of the schema, which proved to be very costly on large schemas.
+#[derive(PartialEq, Eq, Hash, Debug, Clone)]
+pub enum IdentifierType {
+    CheckedCreateInput(Model, Option<RelationField>),
+    CheckedUpdateManyInput(Model),
+    CheckedUpdateOneInput(Model, Option<RelationField>),
+    CompositeCreateEnvelopeInput(CompositeField),
+    CompositeCreateInput(CompositeField),
+    CompositeDeleteManyInput(CompositeField),
+    CompositeUpdateEnvelopeInput(CompositeField),
+    CompositeUpdateInput(CompositeField),
+    CompositeUpdateManyInput(CompositeField),
+    CompositeUpsertObjectInput(CompositeField),
+    CreateManyInput(Model, Option<RelationField>),
+    CreateOneScalarList(ScalarField),
+    Enum(InternalEnum),
+    FieldUpdateOperationsInput(bool, String),
+    Model(Model),
+    NestedUpsertManyInput(RelationField),
+    NestedUpsertOneInput(RelationField),
+    OrderByAggregateInput(ParentContainer, String),
+    OrderByInput(ParentContainer, String),
+    OrderByRelevanceFieldEnum(ParentContainer),
+    OrderByRelevanceInput(ParentContainer),
+    OrderByToManyAggregateInput(ParentContainer),
+    RelationCreateInput(RelationField, bool),
+    RelationUpdateInput(RelationField, bool),
+    ScalarFieldEnum(Model),
+    ScalarFilterInput(Model, bool),
+    ScalarListFilterInput(ScalarField),
+    ScalarListUpdateInput(ScalarField),
+    ToManyCompositeFilterInput(CompositeField),
+    ToManyRelationFilterInput(RelationField),
+    ToOneCompositeFilterInput(CompositeField),
+    ToOneRelationFilterInput(RelationField),
+    TransactionIsolationLevel,
+    UncheckedCreateInput(Model, Option<RelationField>),
+    UncheckedUpdateManyInput(Model, Option<RelationField>),
+    UncheckedUpdateOneInput(Model, Option<RelationField>),
+    UpdateManyWhereCombinationInput(RelationField),
+    UpdateOneWhereCombinationInput(RelationField),
+    UpdateToOneRelWhereCombinationInput(RelationField),
+    WhereInput(ParentContainer),
+    WhereUniqueInput(Model),
+    // Raw string identifier. Used when deferring the allocation + formatting is not worth it.
+    Raw(String),
+}
```

Comment:
My bias here would have been storing identifiers instead of the zippers, but of course that would prevent going through a `Display` impl on this type. This is probably the easiest now, but I really think that long term, we'll have an easier life without the Arc's in every corner of every data structure.